src/shell.js

```diff
@@ -90,6 +92,10 @@ var quit_ = (status, toThrow) => {
 // Determine the runtime environment we are in. You can customize this by
 // setting the ENVIRONMENT setting at compile time (see settings.js).
 
+#if AUDIO_WORKLET
+var ENVIRONMENT_IS_AUDIO_WORKLET = typeof AudioWorkletGlobalScope !== 'undefined';
+#endif
```

Comment:
Maybe it makes sense to make this new env status be more like `ENVIRONMENT_IS_PTHREAD`? That is,  parallel to the main environments, and not one of a disjoint set of the main states. Here is that code:

https://github.com/emscripten-core/emscripten/blob/008017bde42a4b891b0fa0175a1dbe000df99caf/src/shell.js#L120-L126

If this is meant to be a new main environment then arguably it should go into the docs in `var ENVIRONMENT` in settings.js, in particular. But I think it's more like a pthread, a kind of internal state?

Reply:
`ENVIRONMENT_IS_AUDIO_WORKLET` is internal, but it can be useful for JS lib developers to access, to get an outlined short hand to be able to avoid having to type `typeof AudioWorkletGlobalScope !== 'undefined'` out in full in their own detection code.

I am not sure if it would make sense to have that be part of `ENVIRONMENT`, since it would break the [Single Source of Truth](https://en.wikipedia.org/wiki/Single_source_of_truth) rule. I.e. what would it mean to build with a mismatched

`-sAUDIO_WORKLET -sENVIRONMENT=web`
or
`-sENVIRONMENT=web,audioworklet`
command line?

With Web Workers I can see it being different that someone might want to build with pthreads disabled, but with `-sENVIRONMENT=worker` if they plan to run the whole code in a worker. But a similar "I'll do everything manually" deployment cannot happen (is not supported) with Wasm Audio Worklets.

Reply:
I think we might be saying the same thing. I agree that `ENVIRONMENT_IS_AUDIO_WORKLET` is useful for JS devs.

What I'm unsure about is line 117, below - we use that in the same way e.g. `ENVIRONMENT_IS_NODE`. Note that `ENVIRONMENT_IS_PTHREAD` is not present on line 117. I am asking whether we can treat `ENVIRONMENT_IS_AUDIO_WORKLET` more like `ENVIRONMENT_IS_PTHREAD`.

I guess it doesn't matter since audio worklets and `SHELL` have no overlap, but I think it would be simpler to have `ENVIRONMENT_IS_AUDIO_WORKLET` be like `ENVIRONMENT_IS_PTHREAD`.

Reply:
Hmm, line 117 reads:
```
var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER && !ENVIRONMENT_IS_AUDIO_WORKLET;
```
> Note that ENVIRONMENT_IS_PTHREAD is not present on line 117.

The reason that `ENVIRONMENT_IS_PTHREAD` is not present on that line is that `ENVIRONMENT_IS_PTHREAD` implies `ENVIRONMENT_IS_WORKER`, so testing `!ENVIRONMENT_IS_PTHREAD` there would be redundant.

I think it would be incorrect to omit `&& !ENVIRONMENT_IS_AUDIO_WORKLET` on that check. I am thinking maybe someone would like to do a build that would work simultaneously on node.js and with audio worklets enabled, but they might not just ever initialize any audio worklets when they run against node.js. In such case, if `&& !ENVIRONMENT_IS_AUDIO_WORKLET` was omitted there, then the Audio Worklet thread would incorrectly get `ENVIRONMENT_IS_SHELL` be true and likely execute code that will later fail.

If I now locally test just omitting `&& !ENVIRONMENT_IS_AUDIO_WORKLET` from that check, I get test suite errors with
```
RuntimeError: Aborted(Assertion failed: shell environment detected but not enabled at build time.  Add 'shell' to `-sENVIRONMENT` to enable.)
```
so we would then have to enforce that people don't simultaneously target shell and audio worklets at the same time? I think that might be convenient to have, say, I might have a game JS file that would know to run in headless mode in node.js (e.g. for unit tests or a server), but then if the same js file is run in a browser, it would know to pick up webgl rendering and audio worklet enabled audio.

Reply:
I see now, thanks. Makes sense!